driver-sync/src/main/com/mongodb/client/MongoDatabase.java

```diff
@@ -240,8 +240,9 @@
      * Gets the names of all the collections in this database.
      *
      * @return an iterable containing all the names of all the collections in this database
+     * @mongodb.driver.manual reference/command/listCollections listCollections
      */
-    MongoIterable<String> listCollectionNames();
+    ListCollectionsIterable<String> listCollectionNames();
```

Comment:
Returning a more specific type from interface methods is a breaking change for implementors of the interface (the same happens in the `com.mongodb.reactivestreams.client.MongoDatabase` interface). In order to not break semantic versioning we must postpone this feature until version 4.5. What do you think?

Reply:
It's not just breaking for implementors, but also for consumers, as it requires recompilation.

Reply:
Thanks!, you are right. This thread have useful links https://stackoverflow.com/questions/47476509/can-i-change-a-return-type-to-be-a-strict-subtype-and-retain-binary-compatibilit, including quotes from the JLS.

I don't want to contaminate `MongoDatabase` interfaces by adding weird methods, because that would make the API less consistent. Returning a more specific type on the opposite makes the existing API more consistent (don't know why it was not done in the first place, but it causes the problem).

The only ways to tackle the problem I see are

- postpone to 5.0 (I prefer this one);
- introduce a method like `listCollectionNamesConfigurable` that returns `ListCollectionsIterable<String>`, which is ugly.

Reply:
We could also add an overload, but that would make it the only non-fluent method that returns an Iterable.

Given that this was added all the way back in MongoDB 4.0 and no one has yet asked for the feature in the Java driver, I'm also ok with deferring to 5.0.

Reply:
Users can always use `listCollections` and map the results to return the names

Reply:
They can but in that case, the `nameOnly` flag is not set, which means that `authorizedCollections` is ignored. That's why `listCollectionNames` (or its overload, or a similar new method - both alternatives are mentioned in previous comments) is the only way to actually use `authorizedCollections`.

Reply:
I guess, it may also possible to expose `nameOnly` on `ListCollectionsIterable`, thus making the `listCollections` method of full value, while leaving `listCollectionNames` to be a shortcut for a very specific case. But then again, I don't know what I will encounter when trying to so, after all, this was not done when `nameOnly` was introduced.

Reply:
According to the spec the server will ignore nameOnly prior to its addition in 4.0. Which is lucky, also the spec says:

> Drivers MAY allow the nameOnly and authorizedCollections options to be passed when executing the listCollections command for this method.

However, it requires more work to match the spec - checking filters etc. Also it might be surprising to users that one approach is configurable and another isn't.  So deferring to 5.0 seems best.